### PR TITLE
[SMTChecker] Report internal calls in CHC transaction trace

### PR DESCRIPTION
**Pull request by leonardoalt**

### CODE REVIEW

**Comment by stackenbotten** [comment]
There was an error when running `chk_coding_style` for commit `09464c39a29a25f0a1f0327462c950986893806f`:
```
Error: Trailing whitespace found:
 libsolidity/formal/CHC.cpp:1221: addRule(smtutil::Expression::implies(from, to), to.name); 

```
Please check that your changes are working as intended.

File: libsolidity/formal/CHC.cpp [hunk 1]
```diff
@@ -1208,7 +1209,19 @@ smtutil::Expression CHC::predicate(FunctionCall const& _funCall)
...
+	Predicate const& summary = *m_summaries.at(calledContract).at(function);
+	auto from = smt::function(summary, calledContract, m_context);
+	Predicate const& callPredicate = *createSymbolicBlock(
+		functionSort(*function, calledContract, state()),
+		"summary_" + uniquePrefix() + "_call_" + predicateName(function, calledContract),
+		PredicateType::InternalCall,
+		function
+	);
+	auto to = smt::function(callPredicate, calledContract, m_context);
+	addRule(smtutil::Expression::implies(from, to), to.name);
```
Comment (stackenbotten) [inline libsolidity/formal/CHC.cpp]: Coding style error

File: libsolidity/formal/CHC.cpp [hunk 2]
```diff
@@ -1208,7 +1210,17 @@ smtutil::Expression CHC::predicate(FunctionCall const& _funCall)
...
-	return (*m_summaries.at(calledContract).at(function))(args);
+	Predicate const& summary = *m_summaries.at(calledContract).at(function);
+	auto from = smt::function(summary, calledContract, m_context);
+	Predicate const& callPredicate = *createSummaryBlock(
+		*function,
+		*calledContract,
+		kind == FunctionType::Kind::Internal ? PredicateType::InternalCall : PredicateType::ExternalCall
+	);
+	auto to = smt::function(callPredicate, calledContract, m_context);
+	addRule(smtutil::Expression::implies(from, to), to.name);
```
Comment (stackenbotten) [inline libsolidity/formal/CHC.cpp]: Coding style error

**Comment by stackenbotten** [comment]
There was an error when running `chk_coding_style` for commit `32af6dc2f489b763b25ce89392c10017398f5c58`:
```
Error: Trailing whitespace found:
 libsolidity/formal/CHC.cpp:1221: addRule(smtutil::Expression::implies(from, to), to.name); 

```
Please check that your changes are working as intended.

**Review by ekpyron — COMMENTED** [review]
How do you handle free functions in SMT so far? Are they unsupported and abort early or not? If not, is there a test case? (Just to know if I need to keep them in mind when thinking through this...)

**Comment by leonardoalt** [comment]
Yes, they're unsupported and analysis is not performed if free functions or constants are found. There are tests for that under `smtCheckerTests/free_functions`

File: libsolidity/formal/CHC.cpp [hunk 3]
```diff
@@ -1512,16 +1543,27 @@ map<unsigned, vector<unsigned>> CHC::summaryCalls(CHCSolverInterface::CexGraph c
...
 			calls[root].push_back(node);
 			root = node;
 		}
-		for (auto v: _graph.edges.at(node))
-			_addChild({v, root});
-	});
+		auto const& edges = _graph.edges.at(node);
+		for (unsigned v: set<unsigned, decltype(compare)>(begin(edges), end(edges), compare))
+			q.push({v, root});
+	}
```
Comment (blishko) [inline libsolidity/formal/CHC.cpp:1608]: So, what was the result of the discussion? :)
It is not possible to use one of the search utilities?

File: libsolidity/formal/CHC.cpp [hunk 4]
```diff
@@ -1487,10 +1500,12 @@ optional<string> CHC::generateCounterexample(CHCSolverInterface::CexGraph const&
...
 				auto inValues = summaryPredicate->summaryPostInputValues(summaryArgs);
 				auto const& inParams = calledFun->parameters();
-				localState += formatVariableModel(inParams, inValues, "\n") + "\n";
+				if (auto inStr = formatVariableModel(inParams, inValues, "\n"); !inStr.empty())
+					localState += inStr + "\n";
 				auto outValues = summaryPredicate->summaryPostOutputValues(summaryArgs);
 				auto const& outParams = calledFun->returnParameters();
-				localState += formatVariableModel(outParams, outValues, "\n") + "\n";
+				if (auto outStr = formatVariableModel(outParams, outValues, "\n"); !outStr.empty())
+					localState += outStr + "\n";
```
Comment (blishko) [inline libsolidity/formal/CHC.cpp:1550]: These `if`s are responsible for the 350 or so examples changed where only newlines were removed, right? :)

**Review by blishko — DISMISSED** [review]
LGTM!

File: libsolidity/formal/CHC.cpp (hunk 4 above)
Comment (leonardoalt) [reply libsolidity/formal/CHC.cpp:1550]: Yep, there were too many useless blank lines

File: libsolidity/formal/CHC.cpp (hunk 3 above)
Comment (leonardoalt) [reply libsolidity/formal/CHC.cpp:1608]: Yea, here actually not. We *do* need a BFS, but we also need to visit certain nodes multiple times, which happens sometimes with loops.

File: libsolidity/formal/CHC.h [hunk 5]
```diff
@@ -146,7 +146,7 @@ class CHC: public SMTEncoder
 	Predicate const* createBlock(ASTNode const* _node, PredicateType _predType, std::string const& _prefix = "");
 	/// Creates a call block for the given function _function from contract _contract.
 	/// The contract is needed here because of inheritance.
-	Predicate const* createSummaryBlock(FunctionDefinition const& _function, ContractDefinition const& _contract);
+	Predicate const* createSummaryBlock(FunctionDefinition const& _function, ContractDefinition const& _contract, PredicateType _type = PredicateType::FunctionSummary);
```
Comment (mijovic) [inline libsolidity/formal/CHC.h]: Can you document `_type` as well?

File: libsolidity/formal/CHC.cpp [hunk 6]
```diff
@@ -1502,7 +1517,23 @@ optional<string> CHC::generateCounterexample(CHCSolverInterface::CexGraph const&
 				path.emplace_back("State: " + modelMsg);
 		}
 
-		path.emplace_back(txCex);
+		string txCex = summaryPredicate->formatSummaryCall(summaryArgs);
+
+		list<string> calls;
+		auto dfs = [&](unsigned node, unsigned depth, auto&& _dfs) -> void {
```
Comment (mijovic) [inline libsolidity/formal/CHC.cpp:1565]: Why did you need `_dfs` argument here?

File: libsolidity/formal/CHC.h (hunk 5 above)
Comment (leonardoalt) [reply libsolidity/formal/CHC.h]: Done

File: libsolidity/formal/CHC.cpp (hunk 6 above)
Comment (leonardoalt) [reply libsolidity/formal/CHC.cpp:1565]: Because the recursive lambda won't be well typed unless you do that, afaik. I might also be missing something.

**Comment by leonardoalt** [comment]
Updated & rebased

**Comment by leonardoalt** [comment]
@mijovic can you take another look?

**Review by mijovic — APPROVED** [review]
Looks good.
